### ISSUE TITLE
API responding with 500

### ISSUE THREAD

**Issue by manojgaonkar8252** [roskart/dropwizard-jaxws#27]
When i am trying to hit the endpoint [http://localhost:8080/soap/simple](url), i get the below error.

`WARN  [2020-07-08 02:44:05,427] org.apache.cxf.phase.PhaseInterceptorChain: Interceptor for {http://ws.example.jaxws.dropwizard.roskart.com/}SimpleServiceService has thrown exception, unwinding now
! java.lang.NullPointerException: null
! at com.roskart.dropwizard.jaxws.AbstractInvoker.getTargetMethod(AbstractInvoker.java:24)
! at com.roskart.dropwizard.jaxws.InstrumentedInvokers$MeteredInvoker.invoke(InstrumentedInvokers.java:64)
! at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at java.util.concurrent.FutureTask.run(FutureTask.java:266)
! at org.apache.cxf.interceptor.ServiceInvokerInterceptor$2.run(ServiceInvokerInterceptor.java:126)
! at org.apache.cxf.workqueue.SynchronousExecutor.execute(SynchronousExecutor.java:37)
! at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:131)
! at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
! at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
! at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267)
! at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
! at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
! at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
! at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)
! at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:301)
! at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doGet(AbstractHTTPServlet.java:225)
! at javax.servlet.http.HttpServlet.service(HttpServlet.java:503)
! at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:276)
! at io.dropwizard.jetty.NonblockingServletHolder.handle(NonblockingServletHolder.java:50)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1631)
! at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:35)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
! at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:47)
! at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:41)
! at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
! at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:549)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
! at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1363)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
! at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:489)
! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
! at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1278)
! at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
! at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:249)
! at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
! at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:767)
! at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:54)
! at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
! at org.eclipse.jetty.server.Server.handle(Server.java:500)
! at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
! at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
! at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
! at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
! at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
! at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
! at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
! at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
! at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
! at java.lang.Thread.run(Thread.java:748)`

Am i missing something here?

**Comment by roskart** [comment]
This error occurs if you send a GET request to JAX-WS endpoint. You have to POST a SOAP request to JAX-WS endpoint, for example:

```
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ws="http://ws.example.jaxws.dropwizard.roskart.com/">
   <soapenv:Header/>
   <soapenv:Body>
      <ws:echo>
         <arg0>test</arg0>
      </ws:echo>
   </soapenv:Body>
</soapenv:Envelope>
```

**Comment by roskart** [comment]
I'm closing this ticket - this is not an error in dropwizard-jaxws. Feel free to reopen the ticket if necessary.

**Comment by chris--** [comment]
@roskart I have a JAX-WS RI client, that tries a GET request on the endpoint to download the wsdl, leading to the NPE shown above. It retries by adding "?wsdl" as query param and succeeds, so it is not a big deal for me. IMO, I would expect a 404 instead of a 500 and no NPE.

**Comment by roskart** [comment]
@chris-- : Hello. While I agree that NPE is unsuitable, returning 404 is not as straightforward as it sounds. 

If NPE is handled in dropwizard-jaxws and such request is forwarded into CXF stack, CXF also returns status 500 with fault string "No binding operation info while invoking unknown method with params unknown." This is caused by an exception that is thrown in org.apache.cxf.service.invoker.AbstractInvoker. 

Note that WS-I Basic Profile requires that "500 Internal Server Error" HTTP status code is returned if the response envelope is a Fault. This means this shall be handled in servlet filter before the request hits the CXFServlet.

CXFServlet used by dropwizard-jaxws accept HTTP GET requests in two ways:
* request to servlet context root (/soap in example application) returns service list and links to wsdl
* requests for wsdl files (containing ?wsdl) are handled by WSDLGetInterceptor which checks for wsdl and xsd query string parameters

Different containers handle this differently:
* Wildfly returns HTTP status 405 (method not allowed)
* Openliberty returns 200 and outputs name of the service
* Glassfish returns 200 and outputs name of the service and a link to the wsdl

Bottom line:
* returning SOAP Fault and status different than 500 is not acceptable
* handling this in servlet filter is tricky as it would impact some other use cases, such as legacy CXF URIMappingInterceptor that enables using GET requests to invoke SOAP services

If  there is a situation where this issue causes a bug without an easy workaround I am willing to reconsider. For now, I propose that this issue remains closed.
